ding/policy/decision_transformer.py

```diff
@@ -0,0 +1,441 @@
+            os.makedirs(self.log_dir)
+
+        # training and evaluation device
+        self.device = torch.device(self._cfg.device)
+        device = torch.device(self._cfg.device)
+
+        self.start_time = datetime.now().replace(microsecond=0)
+        self.start_time_str = self.start_time.strftime("%y-%m-%d-%H-%M-%S")
+
+        # prefix = "dt_" + env_d4rl_name
+        self.prefix = "dt_" + self.env_name
+
+        save_model_name = self.prefix + "_model_" + self.start_time_str + ".pt"
+        self.save_model_path = os.path.join(self.log_dir, save_model_name)
+        self.save_best_model_path = self.save_model_path[:-3] + "_best.pt"
+
+        log_csv_name = self.prefix + "_log_" + self.start_time_str + ".csv"
+        log_csv_path = os.path.join(self.log_dir, log_csv_name)
+
+        # self.csv_writer = csv.writer(open(log_csv_path, 'a', 1))
+        # csv_header = (["duration", "num_updates", "action_loss",
+        #             "eval_avg_reward", "eval_avg_ep_len", "eval_d4rl_score"])
+
+        # self.csv_writer.writerow(csv_header)
+
+        dataset_path = self._cfg.learn.dataset_path
+        print("=" * 60)
+        print("start time: " + self.start_time_str)
+        print("=" * 60)
+
+        print("device set to: " + str(device))
+        print("dataset path: " + dataset_path)
+        print("model save path: " + self.save_model_path)
+        print("log csv save path: " + log_csv_path)
+
+        self._env = gym.make(self.env_name)
+
+        self.state_dim = self._cfg.model.state_dim
+        self.act_dim = self._cfg.model.act_dim
+
+        self._learn_model = self._model
+        self._optimizer = torch.optim.AdamW(self._learn_model.parameters(), lr=lr, weight_decay=wt_decay)
+
+        self._scheduler = torch.optim.lr_scheduler.LambdaLR(
+            self._optimizer, lambda steps: min((steps + 1) / warmup_steps, 1)
+        )
+
+        self.max_d4rl_score = -1.0
+        self.max_env_score = -1.0
+
+        self.total_updates = 0
+
+    def _forward_learn(self, data: dict) -> Dict[str, Any]:
+        r"""
+            Overview:
+                Forward and backward function of learn mode.
+            Arguments:
+                - data (:obj:`dict`): Dict type data, including at least ['obs', 'action', 'reward', 'next_obs']
+            Returns:
+                - info_dict (:obj:`Dict[str, Any]`): Including current lr and loss.
+        """
+        data_iter = data['data_iter']
+        traj_data_loader = data['traj_data_loader']
+
+        self.log_action_losses = []
+        self._learn_model.train()
+        self._learn_model.to(self.device)  #TODO(pu)
+
+        for _ in range(self.num_updates_per_iter):
+            try:
+                timesteps, states, actions, returns_to_go, traj_mask = next(data_iter)
+            except StopIteration:
+                data_iter = iter(traj_data_loader)
+                timesteps, states, actions, returns_to_go, traj_mask = next(data_iter)
+
```

Comment:
why we use mean and std to norm state here

Reply:
normalization process has been done in dataset.py